elenbaskakova-DOCSUP-728 (#84)
BayoNet/ClickHouse: docs/en/query_language/agg_functions/reference.md

A hash table is used as the algorithm. Because of this, if the passed values ​​are frequently repeated, the function consumes less RAM than `quantileExact`. You can use this function instead of `quantileExact` and specify the weight as 1.

## quantileTDigest(level)(x) {#quantiletdigest}

Approximates the quantile level using the [t-digest](https://github.com/tdunning/t-digest/blob/master/docs/t-digest-paper/histo.pdf) algorithm. The maximum error is 1%. Memory consumption by State is proportional to the logarithm of the number of passed values.

The performance of the function is lower than for `quantile` or `quantileTiming`. In terms of the ratio of State size to precision, this function is much better than `quantile`.

The result depends on the order of running the query, and is nondeterministic.

## median {#median}

Computes [median](https://en.wikipedia.org/wiki/Median) of a numeric data sample.

> 

Reviewer: You have deleted too much :-). I've told only about "This algorithm provides very low accuracy", because I merged this sentece to the previous paragraph in previous commentary.

**Syntax** 

```sql
median(x);
```

> Median is an alias of [quantile(0.5)(x)](#agg_function-quantile)

Reviewer: Point at the end of a sentence.

**Parameters** 

- `x` — Number. Expression resulting in numeric [data types](../../data_types/index.md#data_types), [Date](../../data_types/date.md#date) and [DateTime](../../data_types/datetime.md).

**Returned value**

- The middle value for the sample with odd number of values.
- Arithmetic mean of two middle values for the set with an even number of values.  The function doesn't round the result.  

Type: The same data type as the type of the input data.

**Example**

Input table:

```text
┌─val─┐
│   1 │
│   1 │
│   2 │
│   3 │
└─────┘
```

Query:

```sql
SELECT median(val) FROM table
```

Result:

```text
┌─median(val)─┐
│         1.5 │
└─────────────┘
```

**See Also**

- [quantile](#agg_function-quantile)

## medianDeterministic {#medianDeterministic}

> Computes median or 2-quantile of numeric data set. This function uses [reservoir sampling](https://en.wikipedia.org/wiki/Reservoir_sampling) with a reservoir size up to 8192. 

Reviewer: 2-quantile again.

This algorithm provides very low accuracy

**Syntax** 

```sql
medianDeterministic(x, determinator);
```

MedianDeterministic is an alias of [quantileDeterministiquantile(0.5)(x)](#agg_function-quantiledeterministic)

**Parameters** 

- `x` — Number. Expression resulting in numeric [data types](../../data_types/index.md#data_types), [Date](../../data_types/date.md#date) and [DateTime](../../data_types/datetime.md).
- `determinator` — Number whose hash is used instead of a random number generator in the reservoir sampling algorithm. For the function to work correctly, the same determinator value should not occur too often. For the determinator, you can use an event ID, user ID, and so on. Using integer (0 or greater).
Function is deterministic due to usage of hash of passed number - the "determinator" - in the algorithm. The same determinator value should not occur too often.

**Returned value**

> The middle value will be returned from the set with an odd number of values. For the set with an even number of values arithmetic mean of two middle values will be returned without rounding.  

Reviewer: List as in the `median` description.

Computes the quantile of 'level' exactly. In addition, each value is counted with its weight, as if it is present 'weight' times. The arguments of the function can be considered as histograms, where the value 'x' corresponds to a histogram "column" of the height 'weight', and the function itself can be considered as a summation of histograms.

A hash table is used as the algorithm. Because of this, if the passed values ​​are frequently repeated, the function consumes less RAM than `quantileExact`. You can use this function instead of `quantileExact` and specify the weight as 1.

## quantileTDigest(level)(x) {#quantiletdigest}

Approximates the quantile level using the [t-digest](https://github.com/tdunning/t-digest/blob/master/docs/t-digest-paper/histo.pdf) algorithm. The maximum error is 1%. Memory consumption by State is proportional to the logarithm of the number of passed values.

The performance of the function is lower than for `quantile` or `quantileTiming`. In terms of the ratio of State size to precision, this function is much better than `quantile`.

The result depends on the order of running the query, and is nondeterministic.

## median {#median}

> Computes median or 2-quantile of numeric data set. Median is a middle number of sequence, that separates higher half from the lower half of this sequence. If number of values in the data set is odd then mediana usually is an arithmetic mean of two middle values.

Reviewer: Just "Computes median of a numeric data sample." and link to wiki.  2-quantile or the second quadrile no matter. There is no need to describe median here.

Here and below.

> This function based on [reservoir sampling](https://en.wikipedia.org/wiki/Reservoir_sampling) method with a reservoir size up to 8192. Since this algorithm uses a random number generator this function is non-deterministic.

Reviewer: ```suggestion
This function uses [reservoir sampling](https://en.wikipedia.org/wiki/Reservoir_sampling) method with a reservoir size up to 8192. This algorithm uses a random number generator, thus the function is non-deterministic. Also in has low accuracy. To get exact result, use the [medianExact](#medianexact) function.
```

> This algorithm provides very low accuracy

Reviewer: delete it

**Syntax** 

```sql
median(x);
```

> Alias: `quantile`. 

Reviewer: `quantile` is not the alias, it is another function. If you want to show that median(x)=quantile(0,5)(x), show it in example. For example, you can write something like
```sql
SELECT median(val), quantile(0.5)(val) FROM table
```
and comment this action after the result.

Here and below.

Median is an alias of `quantile` function with an argument of 0.5 (50%).

**Parameters** 

> - `x` — numeric data types, Date and DateTime.

Reviewer: Number. [Expression](link-to-expressions) resulting in numeric [data types](link to introduction to data types), [Date](link to date) and [DateTime](link to datetime). 

Here and below.

**Returned value**

> The middle value will be returned from the set with an odd number of values. For the set with an even number of values arithmetic mean of two middle values will be returned without any rounding.  

Reviewer: ```suggestion
- The middle value for the sample with odd number of values.
- Arithmetic mean of two middle values for the set with an even number of values.  The function doesn't round the result.  
```

Reviewer: Here and below

> Type: Float64, UInt64 and Int64 for numeric, Date and DateTime respectfully for corresponding data types.

Reviewer: Type: The same data type as the type of the input data.

**Example**

Input table:

```text
┌─val─┐
│   1 │
│   1 │
│   2 │
│   3 │
└─────┘
```

Query:

```sql
SELECT median(val) FROM table
```

Result:

```text
┌─median(val)─┐
│         1.5 │
└─────────────┘
```

**See Also**

- [quantile](#agg_function-quantile)

## medianDeterministic {#medianDeterministic}

Computes median or 2-quantile of numeric data set. This function uses [reservoir sampling](https://en.wikipedia.org/wiki/Reservoir_sampling) with a reservoir size up to 8192. 

> Function is deterministic due to usage of hash of passed number - the "determinator" - in the algorithm.

Reviewer: This sentence is too vague for reader who don't know the signature of a function. You can describe the "deterministic" part in Paramenters or even in a separate section. For example, after Returned Values.

└─────────────────────────────┘
```

**See Also** 

- [quantileDeterministic](#agg_function-quantiledeterministic)

## medianExact {#medianexact}

Computes the median exactly. This method uses partially sorted array in the algorithm.
The function consumes O(n) memory, where 'n' is the number of values.

**Syntax** 

```sql
medianExact(x);
```

Alias: `QuantileExact`.

Median is an alias of `QuantileExact` function with an argument of 0.5 (50%).

**Parameters**

- `x` — numeric data types, Date and DateTime.

**Returned value**

The middle value will be returned from the set with an odd number of values. For the set with an even number of values higher value from two middle values will be returned.

Type: Float64, UInt64 and Int64 for numeric, Date and DateTime respectfully for corresponding data types.

**Example**

Query:

```sql
SELECT medianExact(number)
FROM numbers(10)
```

Result:

```text
┌─medianExact(number)─┐
│                   5 │
└─────────────────────┘
```

**See Also**

- [quantileExact](#quantileexact)

## medianExactWeighted {#medianexactweighted}

Computes the median exactly according to the weight of each value. Weight means that value present 'weight' times.

> The arguments of the function can be considered as histograms, where the value 'x' corresponds to a histogram "column" of the height 'weight', 

Reviewer: The last sentence will look better in the Parameters or in the Recycle Bin :-).

and the function itself can be considered as a summation of histograms.

> Since this function is more effective then [medianExact](#medianexact) since it uses hash table in the algorithm.

Reviewer: This function works more efficiently then ... because

**Syntax**

```sql
medianExactWeighted(x, weight);
```

Alias: `quantileExactWeighted`. 

Median is an alias of `quantileExactWeighted` function with an argument of 0.5 (50%).

**Parameters** 

- `x` — numeric data types, Date and DateTime.
- `weight` — shows how many times `x` value repeated. 

**Returned value**

The middle value will be returned from the set with an odd number of values. For the set with an even number of values lower value from two middle values will be returned.

Type: Float64, UInt64 and Int64 for numeric, Date and DateTime respectfully for corresponding data types.

**Example**

Input table:

```text
SELECT *
FROM test1

┌─n─┬─val─┐
│ 0 │   3 │
│ 1 │   2 │
│ 2 │   1 │
│ 5 │   4 │
└───┴─────┘
```

Query:

```sql
SELECT medianExactWeighted(n, val)
FROM test1
```

Result:

```text
┌─medianExactWeighted(n, val)─┐
│                           1 │
└─────────────────────────────┘
```

**See Also** 

- [quantileExactWeighted](#quantileexactweighted)

## medianTiming(x) {#mediantiming}

> Median function optimized for processing of unix timestamps. Computes the median with determined precision.

Reviewer: ```suggestion
Computes the median with determined precision. Optimized for processing of unix timestamps.
```

The result is deterministic. The function is intended for calculating page loading time medians in milliseconds. 

**Syntax** 

```sql
medianTiming(x);
```

Alias: `QuantileTiming`. 

Median is an alias of `QuantileTiming` function with an argument of 0.5 (50%).

**Parameters**

- `x` — numeric data types, Date and DateTime.

**Returned value**

The middle value will be returned from the set with an odd number of values. For the set with an even number of values higher value from two middle values will be returned.

Type: `Float32`.

> **Example**

Reviewer: Try to make an example more corresponding to the goal of the function.
